internal/runtime/hcsv2/process.go

```diff
@@ -114,15 +114,34 @@ func (p *Process) Kill(ctx context.Context, signal syscall.Signal) (err error) {
 		trace.Int64Attribute("pid", int64(p.pid)),
 		trace.Int64Attribute("signal", int64(signal)))
 
-	if err := syscall.Kill(int(p.pid), signal); err != nil {
-		if err == syscall.ESRCH {
-			return gcserr.NewHresultError(gcserr.HrErrNotFound)
-		}
-		return err
-	}
+	// When a container contains more than one process we can fail to unblock
+	// the wait if we only signal the init process. Instead we issue a `runc
+	// kill --all` which then signals all processes in the container.
 	if p.init {
-		p.c.setExitType(signal)
+		if err := p.c.Kill(ctx, signal); err != nil {
+			// Since we use the container logic to kill here it returns a
+			// `gcserr.HrVmcomputeSystemNotFound` but we need a
+			// `gcserr.HrErrNotFound` since we are operating on a process
+			// handle.
+			_, e2 := gcserr.GetHresult(err)
+			if e2 == nil {
+				return gcserr.NewHresultError(gcserr.HrErrNotFound)
+			}
+			// Unknown runc error, return as is.
+			return err
+		}
+	} else {
+		if err := syscall.Kill(int(p.pid), signal); err != nil {
+			if err == syscall.ESRCH {
+				return gcserr.NewHresultError(gcserr.HrErrNotFound)
+			}
+			return err
+		}
+		if p.init {
```

Comment:
It looks like this condition will never occur (since it's inside the `else` block). I think we just want to move the `p.c.setExitType(signal)` to the `if p.init` block?

Reply:
No look at container.Kill it does this already right?

Reply:
Okay but your code here still looks like

```golang
if p.init {
    ...
} else {
    ...
    if p.init {
        ...
    }
}
```

Reply:
Am I crazy? 🤔

Reply:
Oh I should of looked in the full code view. Yes you are right now that we have the first `if/else` we no longer need this block of code. Will fix.